src/alignment.js

```diff
@@ -127,43 +127,36 @@ var clearUnvisitedDOM = function(node) {
   var context = getContext();
   var walker = context.walker;
   var data = getData(node);
-  var keyMap = data.keyMap;
-  var keyMapValid = data.keyMapValid;
-  var lastVisitedChild = data.lastVisitedChild;
-  var child = node.lastChild;
-  var key;
 
-  data.lastVisitedChild = null;
-
-  if (child === lastVisitedChild && keyMapValid) {
+  if (data.attrs[symbols.placeholder] && node !== walker.root) {
     return;
   }
 
-  if (data.attrs[symbols.placeholder] && walker.currentNode !== walker.root) {
-    return;
-  }
-
-  while (child !== lastVisitedChild) {
-    node.removeChild(child);
-    context.markDeleted(/** @type {!Node}*/(child));
+  var keyMap = data.keyMap;
+  var lastVisitedChild = data.lastVisitedChild;
+  var lastChild = node.lastChild;
 
-    key = getData(child).key;
-    if (key) {
-      delete keyMap[key];
+  // Clean the keyMap, removing any unusued keys.
+  if (!data.keyMapValid) {
+    for (var key in keyMap) {
+      var child = keyMap[key];
+      if (child.parentNode !== node) {
+        context.markDeleted(child);
+        delete keyMap[key];
+      }
     }
-    child = node.lastChild;
+    data.keyMapValid = true;
   }
 
-  // Clean the keyMap, removing any unusued keys.
-  for (key in keyMap) {
-    child = keyMap[key];
-    if (!child.parentNode) {
-      context.markDeleted(child);
-      delete keyMap[key];
+  if (lastChild !== lastVisitedChild) {
+    while (lastChild !== lastVisitedChild) {
+      node.removeChild(lastChild);
+      context.markDeleted(/** @type {!Node} **/(lastChild));
```

Comment:
Don't we still need to check if there is a key here to delete? For example if you have items `one`, `two` and `three` and you only visit `one` and `two` in the next pass, `three` will be cleared but it's key won't be until the next time the keymap is marked as invalid.